source/docs/subset_load_balancer.md

```diff
@@ -0,0 +1,273 @@
+encapsulates the filtered `Upstream::HostSet` and `Upstream::LoadBalancer` for a subset.
+
+`ProtobufWkt::Value` is wrapped to provide a cached hash value for the value. Currently,
+`ProtobufWkt::Value` is hashed by first encoding the value as a string and then hashing the
+string. By wrapping it, we can compute the hash value outside the request path for both the
+metadata values provided in `LoadBalancerContext` and those used internally by the SLB.
+
+### Subset Lookup
+
+Currently we require the metadata provided in `LoadBalancerContext` to match a subset exactly in
+order to select the subset for load balancing. Changing this behavior has implications for the
+performance of the subset selection algorithm. The current algorithm, described below, runs in
+`O(N)` time with respect to the number of metadata key-value pairs in the `LoadBalancerContext`.
+
+The metadata key-value pairs from `LoadBalancerContext` must be sorted by key for the algorithm to
+work. Currently we expect lexical order, but the sort order doesn't matter as long as both the
+context and load balancer use the same ordering. Sorting of the `LoadBalancerContext` keys is
+currently handled by `Router::RouteEntryImplBase`.
+
+Given a sequence of N metadata keys and values (previously sorted lexically by key) from
+`LoadBalancerContext`, we can look up the appropriate subset in `O(N)` time as follows. It may be
+helpful to look at the [diagram](#diagram) provided in the example.
+
+1. Initialize `subsets` to refer to the root `LbSubsetMap` and `entry` to point at a null
+   `LbSubsetEntryPtr`.
+2. For each key-value in the metadata:
+   1. Lookup the key in `subsets` to find a `ValueSubsetMap`. (Average constant time.) If not
+      found, exit the loop.
+   2. Lookup the value in the `ValueSubsetMap` to find an `LbSubsetEntry`. (Average constant time.)
+      If not found, exit the loop.
+   3. Assign the `LbSubsetEntry`'s `LbSubsetMap` to `subsets`. (It may be empty.)
+   4. If this is the last key-value pair, assign the `LbSubsetEntry` to `entry`.
+3. If `entry` has been set and has a `Subset` value, we found a matching subset, delegate balancing
+   to the subset's load balancer.
+4. Otherwise, execute the fallback policy.
+
+N.B. `O(N)` complexity presumes that the delegate load balancer executes in constant time.
+
+### Example
+
+Assume a set of hosts from EDS with the following metadata, assigned to a single cluster.
+
+Endpoint | stage | version | type   | xlarge
+---------|-------|---------|--------|-------
+e1       | prod  | 1.0     | std    | true
+e2       | prod  | 1.0     | std    |
+e3       | prod  | 1.1     | std    | true
+e4       | prod  | 1.1     | std    |
+e5       | prod  | 1.0     | bigmem |
+e6       | prod  | 1.1     | bigmem |
+e7       | dev   | 1.2-pre | std    |
+
+Note: Only e1 and e3 have the "xlarge" metadata key.
+
+Given this CDS `envoy::api::v2::Cluster`:
+
+``` json
+{
+  "name": "c1",
+  "lb_policy": "ROUND_ROBIN",
```

Comment:
Do we maintain LB stats for each possible subset? Especially for things like round robin. Same goes for things such as outlier detection (which is needed on per LB pool basis)

Reply:
I think that makes sense, but I haven't done anything with the stats yet.